src/linear_solver.jl

```diff
@@ -0,0 +1,50 @@
+"""
+    AbstractLinearSolver
+
+All linear solvers used within an `ImplicitFunction` must satisfy this interface.
+
+# Required methods
+
+- `presolve(linear_solver, A, y)`: return a matrix-like object `A` for which it is cheaper to solve several linear systems with different vectors `b` (a typical example would be to perform LU factorization).
+- `solve(linear_solver, A, b)`: return a tuple `(x, stats)` where `x` satisfies `Ax = b` and `stats.solved ∈ {true, false}`.
+"""
+abstract type AbstractLinearSolver end
+
+"""
+    IterativeLinearSolver
+
+An implementation of `AbstractLinearSolver` using `Krylov.gmres`.
+"""
+struct IterativeLinearSolver <: AbstractLinearSolver end
```

Comment:
Not for this PR, but this struct could have a field which accepts `A, b` as input and returns a preconditioner :). This function can then be called below in `presolve` which would return both `A` and the preconditioner `P` for consumption in `solve`.

Reply:
For Jacobians, preconditioning `A` using algebraic multi-grid would be really really cool!

Reply:
https://github.com/JuliaLinearAlgebra/Preconditioners.jl

Reply:
Why do a struct with a function field when you can just use pure dispatch?

Reply:
I think this has been the divergence between you and me on this linear solver aspect, and it's funny cause it's really only about style ^^

Reply:
either style is fine, one comes more naturally to me

Reply:
Ok, let's keep the current one for now and we'll see how users react